src/actions/toggleNote.ts

```diff
@@ -1,37 +1,102 @@
 import State from '../@types/State'
 import Thunk from '../@types/Thunk'
+import createThought from '../actions/createThought'
 import deleteAttribute from '../actions/deleteAttribute'
 import setDescendant from '../actions/setDescendant'
 import setNoteFocus from '../actions/setNoteFocus'
 import attribute from '../selectors/attribute'
 import findDescendant from '../selectors/findDescendant'
 import getChildren from '../selectors/getChildren'
+import getNextRank from '../selectors/getNextRank'
 import { registerActionMetadata } from '../util/actionMetadata.registry'
+import createId from '../util/createId'
 import head from '../util/head'
 import reducerFlow from '../util/reducerFlow'
+import resolvePathReference from '../util/resolvePathReference'
 
 /** Toggle the caret between the cursor and its note. Set the selection to the end. If the note is empty, delete it. */
 const toggleNote = (state: State): State => {
   const thoughtId = head(state.cursor!)
   const hasNote = findDescendant(state, thoughtId, '=note')
-  const noteChildren = hasNote ? getChildren(state, hasNote) : []
-  const offset = noteChildren.length ? noteChildren[0].value.length : 0
 
-  return reducerFlow([
-    // create an empty note if it doesn't exist
-    !hasNote
-      ? setDescendant({
-          path: state.cursor!,
-          values: ['=note', ''],
+  // Check if there's a path reference
+  const pathReference = resolvePathReference(state, thoughtId, hasNote)
+
+  // Calculate offset and handle target thought creation if needed
+  let offset = 0
+  let newState = state
```

Comment:
Please do not use mutation for control flow. `reducerFlow` is a better approach to composing reducers.